proposals/0001.md

```diff
@@ -0,0 +1,385 @@
+
+**backtraces**: allows the author to get a back-trace of the failure calls, provided they turn on
+`RUST_BACKTRACE`.
+
+**errors from strings**: allows the author to throw an anonymous error from a string value.
+
+**early returns**: allows the author to return from a function early via a macro call.
+
+**context**: allows the author to add contextual information to a function that can error.
+
+**custom errors**: allows the author to derive the library's custom error type along with special
+`Display` attributions.
+
+`snafu` seems to be the most opionated in terms of types of errors that should be returned. It only
+allows structured errors for a lot of its functionality. It would seem that it makes the most sense
+for library packages but not applications.
+
+`failure` and `anyhow` are very similar in their functionality, if we account for `thiserror` and
+experimental features augmenting `anyhow`.  However, `anyhow` does allow for structured and stringly
+typed errors in its assertion functions.
+
+It is also noteworthy that a library could get enough functionality from `thiserror` while
+applications can build on top of this with another library such as `anyhow` or `failure` to get the
+full functionality of error handling.
+
+# Recommendation
+
+In this section I will recommend what types should be used when encountering error handling in their
+Rust code. Following this I will recommend which packages to use in our libraries and applications.
+
+In [Types of Errors](#types-of-errors) and [`E` is for Error](#e-is-for-error), we outlined
+heuristics that we can follow when encountering error handling situations. To reiterate the
+following rules should be followed when weighing up whether to use `Option` or `Result`:
+
+1. Use `Option` is the failure case is obvious to the caller. This generally means that there is one
+   code path that can't handle the input, or we know the one code path that should be executed. See
+   the examples above to understand these two cases.
+2. Use `Result` when there is more than one failure case in the code path or this function is
+   expected to be composed with more `Result` returning functions. Useful information is expected to
+   be held in the `Err` variant of the `Result`.
+
+Going forward the following recommendations for structuring errors is proposed. An `enum` or a
+`struct` will be created to represent the custom error type. In the case of creating a library,
+these can be left public (`pub`) if they are fine to expose. In the case of opaque errors for which
+the library author wishes to prevent the end-user from doing any control flow, a combination of
+a `pub struct` and hidden fields is recommended. I believe this gives maximal control to the library
+while also preventing the end user from inspecting the error. To demonstrate lets use our previous
+example of a file error but also introduce a deserialization error.
+
+```rust
+pub struct DeserializationError(SomeDeserializationError);
```

Comment:
Hm ok. I _think_ I would in practice have an internal error enum, and unify that at a higher level in a variant `ContentsOfSausage(Box<dyn Error>)`. Because, with this solution you would end up with a lot of standalone structs only for the purpose of hiding the underlying error, or no?

Reply:
Well the struct can hide a private `enum` of opaque errors. So whatever is in the opaque struct is up to you. Whether it's `dyn Error` or something else, all it needs is the `Error` derivation.

My point for having this opaque struct method is to still allow the library to handle the errors if it thinks it can, otherwise it'll eventually become a `String`.

Would you agree?

Reply:
Yeah I agree, but I suspect that I’d be too lazy to exercise this diligently XD.

Reply:
lol, we can always amend ;)